beacon_node/beacon_chain/src/data_availability_checker/availability_view.rs

```diff
@@ -0,0 +1,220 @@
+use crate::data_availability_checker::overflow_lru_cache::PendingComponents;
+use crate::data_availability_checker::ProcessingInfo;
+use crate::AvailabilityPendingExecutedBlock;
+use kzg::KzgCommitment;
+use ssz_types::FixedVector;
+use types::beacon_block_body::KzgCommitments;
+use types::EthSpec;
+
+/// Defines an interface for managing data availability with two key invariants:
+/// 1. Blobs won't be clobbered if we've yet to see the corresponding block.
+/// 2. On block insertion, any non-matching blob commitments are evicted.
+///
+/// Types implementing this trait can be used for validating and managing availability
+/// of blocks and blobs in a cache-like data structure.
+pub trait AvailabilityView<E: EthSpec> {
+    /// The type representing a block in the implementation.
+    type BlockType;
+
+    /// The type representing a blob in the implementation. Must implement `Clone`.
+    type BlobType: Clone;
+
+    /// Checks if a block exists in the cache.
+    ///
+    /// Returns:
+    /// - `true` if a block exists.
+    /// - `false` otherwise.
+    fn block_exists(&self) -> bool;
+
+    /// Checks if a blob exists at the given index in the cache.
+    ///
+    /// Returns:
+    /// - `true` if a blob exists at the given index.
+    /// - `false` otherwise.
+    fn blob_exists(&self, blob_index: u64) -> bool;
+
+    /// Returns the number of blobs that are expected to be present. Returns 0 if we don't have a
+    /// block.
+    ///
+    /// This corresponds to the number of commitments that are present in a block.
+    fn num_expected_blobs(&self) -> usize;
+
+    /// Returns the number of blobs that have been received and are stored in the cache.
+    fn num_received_blobs(&self) -> usize;
+
+    /// Inserts a block into the cache.
+    fn insert_block(&mut self, block: Self::BlockType);
+
+    /// Inserts a blob at a specific index in the cache.
+    ///
+    /// Existing blob at the index will be replaced.
+    fn insert_blob_at_index(&mut self, blob_index: u64, blob: &Self::BlobType);
+
+    /// Converts a blob to its KZG commitment.
+    fn blob_to_commitment(blob: &Self::BlobType) -> KzgCommitment;
+
+    /// Provides mutable access to the underlying blob commitments cache.
+    fn get_cached_blob_commitments_mut(
+        &mut self,
+    ) -> &mut FixedVector<Option<Self::BlobType>, E::MaxBlobsPerBlock>;
+
+    /// Retrieves the KZG commitment of the blob stored at the given index in the block.
+    ///
+    /// Returns `None` if no blob is present at the index.
+    fn get_block_commitment_at_index(&self, blob_index: u64) -> Option<KzgCommitment>;
+
+    /// Merges a given set of blobs into the cache.
+    ///
+    /// Blobs are only inserted if:
+    /// 1. The blob entry at the index is empty and no block exists.
+    /// 2. The block exists and its commitment matches the blob's commitment.
+    fn merge_blobs(&mut self, blobs: FixedVector<Option<Self::BlobType>, E::MaxBlobsPerBlock>) {
```

Comment:
Yea I realized this was broken once I started writing tests for it, came to something similar to this

Reply:
I'm actually getting myself confused, this might have been broken depending on when you commented here, but not for the described reason.

> Suppose we need blobs [0,1,2] and you call this function with blobs [2,3]. Whether or not we have the block, this function will still insert 3. This will cause the is_available() function to break

`[2,3]` here would look like `[None,None,2,3,None,None,None]` so the iteration will always be `MaxBlobsPerBlock` regardless of the `take` because this is a fixed vec

Reply:
It used to look exactly like the code above only without the `max_expected_blobs` line and the `take` on the iterator.

This appears to be fixed now - it was broken before because `get_block_commitment_at_index` didn't differentiate between the commitment not being in the block & the cache not having the block. Thus when it returned `None` you'd hit the `else` and add the unnecessary blob as long as another blob didn't already exist there (which is bad either way).

In the new code, you only hit that `else` if we don't have the block (which is the correct behavior).

Reply:
> This appears to be fixed now - it was broken before because get_block_commitment_at_index didn't differentiate between the commitment not being in the block & the cache not having the block.

Yes i think this was the original issue 👍